pkg/trace/local_client.go

```diff
@@ -0,0 +1,183 @@
+	// is not. Therefore don't create new files after initialization to remain
+	// threadsafe.
+	fileMap map[string]*os.File
+	// canal is a channel for all events that are being written. It acts as an
+	// extra buffer to avoid blocking the caller when writing to files.
+	canal chan Event[Entry]
+}
+
+// NewLocalClient creates a struct that will save all of the events passed to
+// the retuen channel to files based on their "table" (a string field in the
+// event). Each type gets its own file. The internal are purposefully not thread
+// safe to avoid the overhead of locking with each event save. Only pass events
+// to the returned channel. Call CloseAll to close all open files. Goroutine to
+// save events is started in this function.
+func NewLocalClient(cfg *config.Config, logger log.Logger, chainID, nodeID string) (*LocalClient, error) {
+	fm := make(map[string]*os.File)
+	path := path.Join(cfg.RootDir, "data", "traces")
+	for _, table := range splitAndTrimEmpty(cfg.Instrumentation.TracingTables, ",", " ") {
+		fileName := fmt.Sprintf("%s/%s.jsonl", path, table)
+		err := os.MkdirAll(path, 0700)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create directory %s: %w", path, err)
+		}
+		file, err := os.OpenFile(fileName, os.O_APPEND|os.O_CREATE|os.O_RDWR, 0644)
+		if err != nil {
+			return nil, fmt.Errorf("failed to open or create file %s: %w", fileName, err)
+		}
+		fm[table] = file
+	}
+
+	lc := &LocalClient{
+		fileMap: fm,
+		cfg:     cfg,
+		canal:   make(chan Event[Entry], cfg.Instrumentation.TraceBufferSize),
+		chainID: chainID,
+		nodeID:  nodeID,
+		logger:  logger,
+	}
+
+	go lc.draincanal()
+
+	return lc, nil
+}
+
+func (lc *LocalClient) Write(e Entry) {
+	if !lc.IsCollecting(e.Table()) {
+		return
+	}
+	lc.canal <- NewEvent(lc.chainID, lc.nodeID, e.Table(), e)
+}
+
+// ReadTable returns a file for the given table. If the table is not being
+// collected, an error is returned. This method is not thread-safe.
+func (lc *LocalClient) ReadTable(table string) (*os.File, error) {
+	file, has := lc.getFile(table)
+	if !has {
+		return nil, fmt.Errorf("table %s not found", table)
+	}
+
+	return file, nil
+}
+
+func (lc *LocalClient) IsCollecting(table string) bool {
+	if _, has := lc.getFile(table); has {
+		return true
+	}
+	return false
+}
+
+// getFile gets a file for the given type. This method is purposely
+// not thread-safe to avoid the overhead of locking with each event save.
+func (lc *LocalClient) getFile(table string) (*os.File, bool) {
+	f, has := lc.fileMap[table]
+	return f, has
+}
+
+// saveEventToFile marshals an Event into JSON and appends it to a file named after the event's Type.
+func (lc *LocalClient) saveEventToFile(event Event[Entry]) error {
+	file, has := lc.getFile(event.Table)
+	if !has {
+		return fmt.Errorf("table %s not found", event.Table)
+	}
+
+	eventJSON, err := json.Marshal(event)
+	if err != nil {
+		return fmt.Errorf("failed to marshal event: %v", err)
+	}
```

Comment:
thought about using csv, as that should be faster and create significantly smaller files, however the std lib csv is kinda meh tbh and I didn't want to introduce a new dep. if we're ok with using https://github.com/jszwec/csvutil , then I'm totally down.